Mathlib/Data/Finset/Basic.lean

```diff
@@ -812,6 +820,38 @@ theorem cons_swap (hb : b ∉ s) (ha : a ∉ s.cons b hb) :
       ha (mem_cons.mpr (.inl ((mem_cons.mp h).elim symm (fun h ↦ False.elim (hb h))))) :=
   eq_of_veq <| Multiset.cons_swap a b s.val
 
+/-- Split the added element of cons off a Pi type. -/
+@[simps!]
+def consPiProd {s : Finset α} (f : α → Type*) {a : α} (has : a ∉ s)
+    (x : Π i ∈ cons a s has, f i) : f a × Π i ∈ s, f i :=
+  (x a (mem_cons_self a s), fun i hi => x i (mem_cons_of_mem hi))
+
+open Classical in
+/-- Combine a product with a pi type to pi of cons. -/
+noncomputable def prodPiCons {s : Finset α} (f : α → Type*) {a : α} (has : a ∉ s)
+    (x : f a × Π i ∈ s, f i) : (Π i ∈ cons a s has, f i) :=
+  fun i hi =>
+    if h : i = a then cast (congrArg f h.symm) x.1 else x.2 i (mem_of_mem_cons_of_ne has hi h)
+
+/-- The equivalence between pi types on cons and the product. -/
+noncomputable def consPiEquiv {s : Finset α} (f : α → Type*) {a : α} (has : a ∉ s) :
```

Comment:
It's a little strange to be using `Finset.cons` but also to be noncomputable. What do you think of giving this in terms of `Finset.insert` instead, together with decidable equality?